geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java

```diff
@@ -0,0 +1,156 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.Protocol;
+
+import org.apache.geode.redis.RedisIntegrationTest;
+import org.apache.geode.redis.mocks.MockSubscriber;
+import org.apache.geode.test.awaitility.GeodeAwaitility;
+
+public abstract class AbstractSubCommandsIntegrationTest implements RedisIntegrationTest {
+  private Jedis subscriber;
+  private Jedis introspector;
+  private MockSubscriber mockSubscriber;
+
+  @Before
+  public void setup() {
+    mockSubscriber = new MockSubscriber();
+    subscriber = new Jedis("localhost", getPort());
+    introspector = new Jedis("localhost", getPort());
+  }
+
+  @After
+  public void teardown() {
+    if (mockSubscriber.getSubscribedChannels() > 0) {
+      mockSubscriber.unsubscribe();
+    }
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 0);
+  }
+
+  @Test
+  public void Pubsub_shouldError_givenTooFewArguments() {
+    assertAtLeastNArgs(introspector, Protocol.Command.PUBSUB, 1);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void pubsub_shouldReturnError_givenUnknownSubcommand() {
+    String expected = String.format(UNKNOWN_SUBCOMMAND_ERROR_BASE, "nonesuch");
+
+    assertThatThrownBy(() -> introspector.sendCommand(Protocol.Command.PUBSUB, "nonesuch"))
+        .hasMessageContaining(expected);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void channels_shouldReturnListOfAllChannels_whenCalledWithoutPattern() {
+    ArrayList<byte[]> expectedChannels = new ArrayList<>();
+    expectedChannels.add("foo".getBytes());
+    expectedChannels.add("bar".getBytes());
+    Runnable runnable =
+        () -> subscriber.subscribe(mockSubscriber, "foo", "bar");
+    Thread subscriberThread = new Thread(runnable);
+
+    subscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 2);
+    ArrayList<byte[]> result = (ArrayList<byte[]>) introspector
+        .sendCommand(Protocol.Command.PUBSUB, "channels");
+
+    assertThat(result).containsExactlyInAnyOrderElementsOf(expectedChannels);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void channels_shouldReturnListOfMatchingChannels_whenCalledWithPattern() {
+    ArrayList<byte[]> expectedChannels = new ArrayList<>();
+    expectedChannels.add("foo".getBytes());
+    Runnable runnable =
+        () -> subscriber.subscribe(mockSubscriber, "foo", "bar");
+    Thread subscriberThread = new Thread(runnable);
+
+    subscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 2);
+    ArrayList<byte[]> result =
+        (ArrayList<byte[]>) introspector
+            .sendCommand(Protocol.Command.PUBSUB, "channels", "fo*");
```

Comment:
For this test as well as the test below it we could use `introspector.pubsubChannels("fo*")`

Reply:
yeah, the compiler still wants the cast, (this time to list<String>), but we don't need to suppress the unchecked cast warning  :)

Reply:
If you replace the type with `List<String>` rather than `ArrayList<String>`, you won't have to cast it.